go/pkg/chunker/chunker.go

```diff
@@ -60,20 +65,74 @@ func NewFromBlob(blob []byte, chunkSize int) *Chunker {
 // any time. This means that late errors due to mismatch of digest and
 // contents are possible.
 func NewFromFile(path string, dg digest.Digest, chunkSize int) *Chunker {
+	c := NewFromReader(reader.NewFileReadSeeker(path, IOBufferSize), dg, chunkSize)
+	c.path = path
+	return c
+}
+
+// NewCompressedFromFile is similar to NewFromFile, except that it'll compress
+// the contents of the file using zstd compression.
+func NewCompressedFromFile(path string, dg digest.Digest, chunkSize int) (*Chunker, error) {
+	r, err := reader.NewCompressedFileSeeker(path, IOBufferSize)
+	if err != nil {
+		return nil, err
+	}
+	c := NewFromReader(r, dg, chunkSize)
+	c.path = path
+	return c, nil
+}
+
+// NewFromReader is similar to NewFromFile, except that it'll take in any
+// arbitrary reader.ReadSeeker. Note that the digest is still for convenience,
+// rather than being actual used by the Chunker.
+func NewFromReader(r reader.ReadSeeker, dg digest.Digest, chunkSize int) *Chunker {
 	if chunkSize < 1 {
 		chunkSize = DefaultChunkSize
 	}
 	if chunkSize > IOBufferSize {
 		chunkSize = IOBufferSize
 	}
 	return &Chunker{
-		r:         reader.NewFileReadSeeker(path, IOBufferSize),
+		r:         r,
 		chunkSize: chunkSize,
 		digest:    dg,
-		path:      path,
 	}
 }
 
+// CompressChunker will, if possible, modify the chunker so it'll return compressed
+// reads instead. This is irreversible.
+func CompressChunker(ch *Chunker) error {
```

Comment:
I wished to do that initially, but the problem on that approach is that there are chunkers passed in from other parts of the codebase (eg [1](bazelbuild/remote-apis-sdks/blob/master/go/pkg/tree/tree.go#L87) and [2](bazelbuild/remote-apis-sdks/blob/master/go/pkg/tool/tool.go#L228)). I think the control for whether to compress or not blobs should be centralized around the CAS, and hidden logic from the rest of the code. This way I'm avoiding duplication and isolating logic.

Alternatively, I could make `UploadIfMissing` receive filepaths instead of chunks, but that also sounds counterproductive as it makes it impossible to mix file/blob/proto chunkers.

Reply:
That's why I suggested to make a single Client wrapper for the Chunker constructor and use it everywhere. That also avoids duplication and centralizes the logic in the CAS, and doesn't even modify the existing code much, but it also has the advantages of having no side effects -- Chunker is not "mutable" after creation, and no functions need to mutate the arguments. Plus it would ensure correctness -- that all Chunker users now use the new constructor -- unlike the current change to WriteBlob only.